File: test/itemicon.test.ts
/**
 * Icon keyword matching.
 *
 *   npm run test:icon
 *
 * The rules that matter: the name beats the other fields, longer phrases beat
 * the words inside them, and a match must be on a word boundary — "wave" must
 * not select the microwave icon.
 */

import { iconKeyFor } from '@/lib/itemIcon';

let failures = 0;

function check(label: string, ok: boolean, detail = '') {
  if (!ok) failures++;
  console.log(`${ok ? 'PASS' : 'FAIL'}  ${label}${detail ? `  — ${detail}` : ''}`);
}

function expect(subject: Parameters<typeof iconKeyFor>[0], want: string) {
  const got = iconKeyFor(subject);
  const desc = subject.name ?? JSON.stringify(subject);
  check(`${desc} → ${want}`, got === want, got === want ? '' : `got ${got}`);
}

// Straight matches from the name.
expect({ name: 'Bosch Dishwasher' }, 'dishwasher');
expect({ name: 'LG Refrigerator' }, 'fridge');
expect({ name: 'Sony Bravia' }, 'tv');
expect({ name: 'DeWalt Table Saw' }, 'saw');
expect({ name: 'Dyson vacuum' }, 'vacuum');
expect({ name: 'Weber BBQ' }, 'grill');
expect({ name: 'MacBook Pro' }, 'laptop');
expect({ name: 'Nespresso machine' }, 'coffee');

// Longest phrase wins over the words inside it.
expect({ name: 'Washing machine' }, 'washer');
expect({ name: 'Bosch washer' }, 'washer');
expect({ name: 'Combination microwave' }, 'microwave');
expect({ name: 'Air conditioner' }, 'aircon');
expect({ name: 'Water heater' }, 'boiler');

// "table saw" must not become a table; "table" alone still should.
expect({ name: 'Table saw' }, 'saw');
expect({ name: 'Dining table' }, 'table');

// Word boundaries: no matching inside a longer word.
expect({ name: 'Wavelength meter' }, 'box');
expect({ name: 'Carpet cleaner' }, 'box');
expect({ name: 'Scartissue' }, 'box');

// Case and punctuation.
expect({ name: 'SAMSUNG TV' }, 'tv');
expect({ name: 'e-bike, blue' }, 'bike');

// Falls through name → other fields → category → generic.
expect({ name: 'SHXM4AY55N', notes: 'the dishwasher in the kitchen' }, 'dishwasher');
expect({ name: 'Unknown thing', category: 'vehicle' }, 'car');
expect({ name: 'Unknown thing', category: 'furniture' }, 'sofa');
expect({ name: 'Unknown thing' }, 'box');
expect({}, 'box');

// The name outranks the other fields even when both match.
expect({ name: 'Garden mower', notes: 'stored next to the fridge' }, 'mower');

console.log(failures === 0 ? '\nall green' : `\n${failures} failure(s)`);
process.exit(failures === 0 ? 0 : 1);

File: test/item_icon_test.dart
/// Choosing an icon for an item with no photo.
///
///   dart test test/item_icon_test.dart
///
/// Translated from `test/itemicon.test.ts`. The interesting cases are all about
/// precedence: which field wins, and which keyword wins inside a field.
library;

import 'package:stash_it/logic/item_icon.dart';
import 'package:flutter_test/flutter_test.dart';

IconKey iconFor({String? name, String? brand, String? model, String? notes}) =>
    iconKeyFor(IconSubject(name: name, brand: brand, model: model, notes: notes));

void main() {
  group('the obvious ones', () {
    test('a fridge', () => expect(iconFor(name: 'Fridge'), IconKey.fridge));
    test('a dishwasher', () => expect(iconFor(name: 'Bosch Dishwasher'), IconKey.dishwasher));
    test('a synonym counts', () => expect(iconFor(name: 'Refrigerator'), IconKey.fridge));
    test('and case does not', () => expect(iconFor(name: 'KETTLE'), IconKey.kettle));
  });

  group('the floor', () {
    // A wrong icon is worse than a neutral one: it looks like the app has
    // misunderstood the thing.
    test('nothing recognisable is a box', () {
      expect(iconFor(name: 'Thing'), IconKey.box);
    });

    test('and nothing at all is too', () => expect(iconFor(), IconKey.box));
    test('an empty name is too', () => expect(iconFor(name: ''), IconKey.box));
  });

  group('longest phrase wins', () {
    /*
      THE PRECEDENCE THE INDEX EXISTS FOR. "washing machine" and "washer" are
      different keys, and "machine" appears inside the first. Sorted by length,
      the phrase is tested before either word inside it.
    */
    test('a washing machine is a washer, not whatever "machine" would be', () {
      expect(iconFor(name: 'Washing machine'), IconKey.washer);
    });

    test('a table saw is a saw, not a table', () {
      expect(iconFor(name: 'DeWalt table saw'), IconKey.saw);
    });

    test('a pizza oven is a grill, not an oven', () {
      expect(iconFor(name: 'Ooni pizza oven'), IconKey.grill);
    });

    test('an air fryer is an oven, not a fan', () {
      expect(iconFor(name: 'Air fryer'), IconKey.oven);
    });
  });

  group('word boundaries', () {
    // "microwave" must not match inside "wave", and nothing should match a
    // word it merely happens to be a substring of.
    test('a keyword inside a longer word does not match', () {
      expect(iconFor(name: 'Microwaves are loud'), IconKey.box);
      expect(iconFor(name: 'Carpet'), IconKey.box);
      expect(iconFor(name: 'Sawdust'), IconKey.box);
    });

    test('but a keyword next to punctuation does', () {
      expect(iconFor(name: 'TV, wall-mounted'), IconKey.tv);
      expect(iconFor(name: '(kettle)'), IconKey.kettle);
    });

    /*
      "e-bike" ends in a word character but begins next to a hyphen, and the
      hyphen is why `\b` was not trusted: a phrase whose edge is punctuation
      needs the boundary written out.
    */
    test('and a hyphenated one does', () {
      expect(iconFor(name: 'Specialized e-bike'), IconKey.bike);
    });
  });

  group('which field wins', () {
    /*
      The name gets its own pass first, because it carries the most intent.
      "Bosch SHXM4AY55N" in a model field tells you nothing; "dishwasher" in
      the name tells you everything.
    */
    test('the name beats the notes', () {
      expect(
        iconFor(name: 'Kettle', notes: 'bought at the same time as the sofa'),
        IconKey.kettle,
      );
    });

    test('but the rest is used when the name says nothing', () {
      expect(iconFor(name: 'The big one', notes: 'chest freezer'), IconKey.fridge);
    });

    test('brand and model are searched too', () {
      expect(iconFor(name: 'Spare', model: 'Roomba j7'), IconKey.vacuum);
    });
  });

  test('every key except box has at least one keyword', () {
    for (final key in IconKey.values) {
      final words = keywords[key];
      expect(words, isNotNull, reason: '$key is missing from the keyword table');
      if (key != IconKey.box) {
        expect(words, isNotEmpty, reason: '$key has no keywords');
      }
    }
  });

  test('and no keyword is claimed by two keys', () {
    final seen = <String, IconKey>{};
    keywords.forEach((key, words) {
      for (final w in words) {
        expect(seen.containsKey(w), isFalse,
            reason: '"$w" is claimed by both ${seen[w]} and $key');
        seen[w] = key;
      }
    });
  });
}
